File: core/src/main/java/com/terralite/core/registry/MutableRegistry.java
package com.terralite.core.registry;

public interface MutableRegistry<T> extends Registry<T> {
    T register(ResourceId id, T value);

    default T register(ResourceKey<T> resourceKey, T value) {
        if (!key().equals(resourceKey.registry())) {
            throw new IllegalArgumentException("Resource key belongs to registry " + resourceKey.registry().id() + ", not " + id());
        }
        return register(resourceKey.id(), value);
    }

    FrozenRegistry<T> freeze();
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/MutableRegistry.kt
package com.terralite.core.registry

/** [Registry] that accepts new entries and can be sealed into a [FrozenRegistry]. */
interface MutableRegistry<T> : Registry<T> {

    fun register(id: ResourceId, value: T): T

    fun register(key: ResourceKey<T>, value: T): T {
        require(key().id == key.registry.id) {
            "Resource key belongs to registry ${key.registry.id}, not ${id()}"
        }
        return register(key.id, value)
    }

    fun replace(id: ResourceId, value: T): T

    fun freeze(): FrozenRegistry<T>
}
